### DISCUSSION TITLE
Yalmip is not able to convert the model to a form supported by Gurobi

### DISCUSSION THREAD

**Discussion by zth72** [yalmip/YALMIP#1442] [General discussion] [answered]
![1cc5e13c43ed14fbe23845c01b0c988f](https://github.com/user-attachments/assets/b3f12554-8a3a-4ca3-a005-98fe43a3b437)
Here's the code I wrote. When I run it in matlab, I get the following error:
Warning: Solver not applicable (gurobi does not support signomial equality constraints)

load('BESSSA1.mat');
load('GridA_emd.mat');
BESSN8 = BESSA1(1,:); 
GridN8 = GridA_emd; 
Nday8 = ones(1,288);

%% 超级电容容量配置的相关参数的初始化
S_scmax = sdpvar(1,1);   % 超电容容量
S_sc = sdpvar(1,24*12);  % 超级电容荷电量
E_scmax = sdpvar(1,1);   % 超级电容功率限值

C = [];
C = [C, 0 <= S_scmax <= 15];
C = [C, 0.15*S_scmax <= S_sc, S_sc <= 0.85*S_scmax];
C = [C, E_scmax == max(abs(min(BESSN8)/0.9), 0.9*max(BESSN8))];

Echarge_bess = sdpvar(1,24*12);   % 超级电容充电功率
Edischarge_bess = sdpvar(1,24*12); % 超级电容放电功率
B_sc = binvar(1,24*12);          % 超级电容充放电状态标识位置
M = 1e6;
C = [C, 0 <= Echarge_bess, Echarge_bess <= B_sc * M];
C = [C, 0 <= Edischarge_bess, Edischarge_bess <= (1 - B_sc) * M];
C = [C, 0 <= Echarge_bess, Echarge_bess <= E_scmax];
C = [C, 0 <= Edischarge_bess, Edischarge_bess <= E_scmax];

C = [C, S_sc(1) - 0.9 * Echarge_bess(1) / 12 + Edischarge_bess(1) / 0.9 / 12 == S_sc(end)];
C = [C, S_sc(2:end) == 0.995 * S_sc(1:end-1) + 0.9 * Echarge_bess(2:end) / 12 - Edischarge_bess(2:end) / 0.9 / 12];

% 参数初始化
SOC_min = 0.15;    % 能量型储能设备的最小荷电状态
SOC_max = 0.85;    % 能量型储能设备的最大荷电状态
SOC_0 = 0.5;       % 初始荷电状态
delta_T = 5 * 60;  % 时间间隔 (5分钟，以秒为单位)

% 假设 P_ESS_e 是储能设备的功率指令数据向量 (单位：MW)
P_ESS_e = sdpvar(1, 288);  % 示例数据，实际应替换为真实功率数据
C = [C, P_ESS_e == Echarge_bess - Edischarge_bess];

% 计算累积功率指令
cum_P_ESS_e = sdpvar(1, 288);
C = [C, cum_P_ESS_e*3600 == cumsum(P_ESS_e) * delta_T ];

% 公式 (30) 的计算
S_scmax = max(-(min(cum_P_ESS_e) / (SOC_0 - SOC_min)), max(cum_P_ESS_e) / (SOC_max - SOC_0));

zheng_bodong = sdpvar(1, 24*12);  % 弃电功率(正向波动)
fu_bodong = sdpvar(1, 24*12);     % 负向波动
zhengfu_bodong = sdpvar(1, 24*12); % 正负向波动

%% 修改①：引入基准线变量，去替代0标准线
jizhunxian = sdpvar(1, 1);   
C = [C, min(BESSN8) <= jizhunxian, jizhunxian <= max(BESSN8)];

dita = sdpvar(1, 24*12);    % 波动功率          
sumdita = sdpvar(1, 1);     % 累计波动量          
new_line = sdpvar(1, 24*12); % 平抑后的风电曲线高频分量

%% 修改②：
C = [C, 0 <= zheng_bodong, zheng_bodong <= max(BESSN8, 0)];
C = [C, min(BESSN8, 0) <= fu_bodong, fu_bodong <= 0];
C = [C, zheng_bodong == max(BESSN8, 0) - Echarge_bess];
C = [C, fu_bodong == min(BESSN8, 0) + Edischarge_bess];

C = [C, 0 <= zheng_bodong, zheng_bodong <= max(BESSN8, 0) - jizhunxian];
C = [C, min(BESSN8, 0) - jizhunxian <= fu_bodong, fu_bodong <= 0];
C = [C, zheng_bodong == max(BESSN8, 0) - jizhunxian - Echarge_bess];
C = [C, fu_bodong == min(BESSN8, 0) - jizhunxian + Edischarge_bess];

%% min和max的线性化
zheng_SCN8jizhunxian = sdpvar(1, 24*12);               
fu_SCN8jizhunxian = sdpvar(1, 24*12);                        
Bzhengfu_SCN8jizhunxian = binvar(1, 24*12);             
C = [C, 0 <= zheng_SCN8jizhunxian, zheng_SCN8jizhunxian <= M * Bzhengfu_SCN8jizhunxian];           
C = [C, 0 <= fu_SCN8jizhunxian, fu_SCN8jizhunxian <= M * (1 - Bzhengfu_SCN8jizhunxian)];         
C = [C, zheng_SCN8jizhunxian - fu_SCN8jizhunxian == BESSN8 - jizhunxian];            

C = [C, 0 <= zheng_bodong, zheng_bodong <= zheng_SCN8jizhunxian];
C = [C, -fu_SCN8jizhunxian <= fu_bodong, fu_bodong <= 0];
C = [C, zheng_bodong == zheng_SCN8jizhunxian - Echarge_bess];
C = [C, fu_bodong == -fu_SCN8jizhunxian + Edischarge_bess];   

C = [C, zhengfu_bodong == zheng_bodong + fu_bodong];   
C = [C, sumdita == sum(Nday8 .* (zheng_bodong - fu_bodong))]; 
 C = [C, zhengfu_bodong == fu_bodong];  % 这里正向波动，也可以直接当作弃电功率
% % %% 寿命评估模型
SOC_t=sdpvar(1,24*12);
C=[C, SOC_t*S_scmax==S_sc];
% C=[C,SOC_t>=0.15];
% C=[C,SOC_t<=0.85];
Dcir_t=sdpvar(1,24*12);
C=[C,Dcir_t==1-SOC_t];
% Rcir_t=sdpvar(1,24*12)
% C=[C,Rcir_t==765*exp(1.69*(1-Dcir_t)).*(1./Dcir_t).^0.19];
Rdod_t=sdpvar(1,24*12);
x=sdpvar(1,288);
y=sdpvar(1,288);
C=[C,x==Dcir_t.^-0.19];
u=sdpvar(1,1);
C=[C,-1.69*SOC_t]
C=[C,y==exp(u)];
C=[C,Rdod_t==x.*y];
Yba=sdpvar(1,1);
z=sdpvar(1,1);
C=[C,z==sum(Rdod_t)];
w=sdpvar(1,1)
C=[C, w*z==1];
C=[C,Yba*w==765];
%% 成本分析
% 电池容量配置年成本系数：
c_sc=sdpvar(1,1);
q=sdpvar(1,1);
C=[C,q==1.05^Yba]
C=[C,c_sc* (q - 1)  ==10^7 * 0.05 * q  ];  % = 963元/(度电容量*年)  = 963000元/(MW容量*年)
F_sc = sdpvar(1, 1);  % 超级电容的投资和运行成本
F_sc_P=sdpvar(1,1);
F_sc_S=sdpvar(1,1);
C=[C,  F_sc_P==2*c_sc*E_scmax];
C=[C,F_sc_S==2*c_sc*S_scmax]
C = [C, F_sc == F_sc_S+F_sc_P];
% compensate_value = sdpvar(1, 1);
% C = [C, compensate_value == sum(abs(BESSN8)) / 12];  % 每天的补偿量

% 最大充放电总量系数：
Maxcdiscxishu = 100000;
C = [C, sum(Nday8 .* Echarge_bess) <= Maxcdiscxishu *S_scmax];
ops = sdpsettings('verbose', 2, 'solver', 'gurobi');
ops.cplex.MIPGap = 0.01;
ops.cplex.TimeLimit = 100;

% 进行求解计算
c_comp=sdpvar(1,1);
C=[C,c_comp*12==sumdita * 0.32 * 10000 * 365];
result = optimize(C, c_comp + F_sc*1.02, ops);
if result.problem == 0
fprintf('求解成功！\n');
else
error('求解出错！请查找错误来源');

end

**Comment by johanlofberg** [comment, marked as answer]
This

C=[C,x==Dcir_t.^-0.19];

Is not MILP/MIQP-representable and thus not possible. If you rewrite it using exp/log, and use the develop branch of YALMIP, it will employ the recent support for nonlinear global optimization in Gurobi

**Reply by zth72** [reply]
Thank you for your answer, but exp is based on e as the base to power, log is logarithmic, so how should I write this equation

**Reply by johanlofberg** [reply]
I don't understand the question. exp/log are supported in gurobi 11 if you use the develop branch of YALMIP, and will lead to gurobi solving the global nonlinear optimization problem through spatial branch and bound
